recipes/automake/all/conanfile.py

```diff
@@ -34,11 +34,11 @@ def configure(self):
         del self.settings.compiler.cppstd
         del self.settings.compiler.libcxx
 
-    def requirements(self):
-        self.requires("autoconf/2.71")
-        # automake requires perl-Thread-Queue package
+    #def requirements(self):
+    #    automake requires perl-Thread-Queue package
```

Comment:
I would do it in reverse:
- always add `autoconf` as a requirement
- and only add `autoconf as a build requirement when cross building

This is needed because when adding `automake` as a requirement/build requirement, autoconf is always required as well.

```suggestion
    def requirements(self):
        self.requires("autoconf/2.71")
        # automake requires perl-Thread-Queue package
```

Reply:
But `autoconf` is not really a requirement. Are you proposing abusing the system to avoid listing both packages as `build_requirements` on many recipes?
Also, I'm not crosscompiling. I'm using the 2 profiles approach but it is actually the same profile twice. This is because on our CI pipeline we specify the `build` profile once, and then for each package we list the desired `host` profile. When we want to build an x86_64 package we end up using the same profile as build and host, E.g.
```
conan create someLib/someVersion@user/channel -pr:b x86_64_gcc8 -pr:h x86_64_gcc8
```

But yes, I noticed that changing `autoconf` to be a `build_req` has consequences on other recipes. For example to build `libtool` I also changed `automake` to be a `build_requirement` and then had to add `m4` as well, since we're no longer inheriting the pacakges.
Even though it is more tedious, I think it is the right way.

Reply:
For consumers of automake, it is a plugin for autoconf when regenerating `configure` and only needed then.
When only needing some scripts (e.g. the compile / ar_lib scripts), then autoconf is unneeded.

If I remember correctly, we isolated the dependency on autoconf to this location here for 2 reasons:
- It's much easier to bump autoconf: only do it here and be done with it.
Also, I'm not sure how an automake built with autoconf/2.73 would behave when combined with an older autoconf/2.71 (=recipe where autoconf has not been bumped yet).
- I also remember a problem when trying to build multiple packages in a dependency tree.
Conan choked when 2 recipes with different versions of autoconf /automake/libtool were present in the same tree.
I don't know whether conan behaves better now. (@SpaceIm  does this problem ring a bell to you?)

Reply:
A practical example, using the 2 matching profiles as I mentioned above + my this version of `automake`, in order to make `patchelf` compile I had to change this 
```
    def build_requirements(self):
        self.build_requires("libtool/2.4.6")
```
With
```
    def build_requirements(self):
        self.build_requires("libtool/2.4.6")
        self.build_requires("automake/1.16.4")
        self.build_requires("autoconf/2.71")
```
Which kind of makes sense. If the `patchelf` needs this packages to build, then they should be marked explicitly as build requirements.

Reply:
>     * I also remember a problem when trying to build multiple packages in a dependency tree.
>       Conan choked when 2 recipes with different versions of autoconf /automake/libtool were present in the same tree.
>       I don't know whether conan behaves better now. (@SpaceIm  does this problem ring a bell to you?)

build requirements versions can't be overridden, so if a conanfile "build requires" libtool/2.4.6 and autoconf/2.69 for example, it fails because libtool drags automake, which requires autoconf/2.71 currently.

Reply:
> 
> 
> > ```
> > * I also remember a problem when trying to build multiple packages in a dependency tree.
> >   Conan choked when 2 recipes with different versions of autoconf /automake/libtool were present in the same tree.
> >   I don't know whether conan behaves better now. (@SpaceIm  does this problem ring a bell to you?)
> > ```
> 
> build requirements versions can't be overridden, so if a conanfile "build requires" libtool/2.4.6 and autoconf/2.69 for example, it fails because libtool drags automake, which requires autoconf/2.71 currently.

Can we avoid this by not adding `autoconf` to the requirements of `automake` and not adding `automake` to the requirements of `libtool`?

Reply:
I think so, but it would break many recipes.

Reply:
I basically added packages until it stopped complaining and built, that's why m4 is not listed.

Reply:
Good point about `m4`.
That dependency would be pulled in when adding autoconf as a build requirement.

@ericriff, I would just keep it in the current state.
Too many recipes would break with no clear gain.

Reply:
I can decline this PR and keep all the fixes I need to myself then. It would be annoying since I try to keep our fork as close as possible to the official index, but  I understand your point.
The only gain to listing the build requirements properly is that it allows builds with pr:h=pr:b.
It is my understanding that the dual profile method will be enforced in Conan 2.0 (but I might be wrong since I don't remember where I got that information from). If that is in deed the case, will it only be enforced to crosscompile? Is it expected to continue using a single profile when doing native builds?

Reply:
By applying https://github.com/conan-io/conan-center-index/pull/8056#discussion_r749529219, the build requirement is only added when cross building.

Reply:
`pr:h=pr:b` is not actually crosscompiling. Both profiles are the same x86_64 profile. Would that work anyways?

Reply:
I meant to say that the build requirement is only added when using conan's 2 profile build approach.

Reply:
Maybe I'm missing something here. 
I changed the code to list the requirements and build_requirements like this
```py
    def requirements(self):
        self.requires("autoconf/2.71")
        # automake requires perl-Thread-Queue package

    def build_requirements(self):
        if hasattr(self, "settings_build"):
            self.build_requires("autoconf/2.71")
        if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
            self.build_requires("msys2/cci.latest")
```
But when I build with two identical profiles the build doesn't see autoconf in PATH
```
$ conan create recipes/automake/all automake/1.16.4@eriff/test -pr:b default -pr:h default
Exporting package recipe
automake/1.16.4@eriff/test exports: File 'conandata.yml' found. Exporting it...
automake/1.16.4@eriff/test exports: Copied 1 '.yml' file: conandata.yml
automake/1.16.4@eriff/test exports_sources: Copied 14 '.patch' files
automake/1.16.4@eriff/test: A new conanfile.py version was exported
automake/1.16.4@eriff/test: Folder: /home/eric/.conan/data/automake/1.16.4/eriff/test/export
automake/1.16.4@eriff/test: Exported revision: 1244ed17c2051671b31c501bab8541ab
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

automake/1.16.4@eriff/test: Forced build from source
automake/1.16.4@eriff/test (test package): Installing package
Requirements
Packages
Build requirements
    autoconf/2.71 from 'conancenter' - Cache
    automake/1.16.4@eriff/test from local cache - Cache
    m4/1.4.19 from 'conancenter' - Cache
Build requirements packages
    autoconf/2.71:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    automake/1.16.4@eriff/test:258f6f05ca54813b3d7180161753df7f9ccf7e1d - Build
    m4/1.4.19:24647d9fe8ec489125dfbae4b3ebefaf7581674c - Cache

Installing (downloading, building) binaries...
m4/1.4.19: Already installed!
m4/1.4.19: Appending PATH environment variable: /home/eric/.conan/data/m4/1.4.19/_/_/package/24647d9fe8ec489125dfbae4b3ebefaf7581674c/bin
m4/1.4.19: Setting M4 environment variable: /home/eric/.conan/data/m4/1.4.19/_/_/package/24647d9fe8ec489125dfbae4b3ebefaf7581674c/bin/m4
autoconf/2.71: Already installed!
autoconf/2.71: Appending PATH env var with : /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin
autoconf/2.71: Setting AC_MACRODIR to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/share/autoconf
autoconf/2.71: Setting AUTOCONF to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/autoconf
autoconf/2.71: Setting AUTORECONF to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/autoreconf
autoconf/2.71: Setting AUTOHEADER to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/autoheader
autoconf/2.71: Setting AUTOM4TE to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/autom4te
autoconf/2.71: Setting AUTOM4TE_PERLLIBDIR to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/share/autoconf
automake/1.16.4@eriff/test: Configuring sources in /home/eric/.conan/data/automake/1.16.4/eriff/test/source
Downloading automake-1.16.4.tar.gz completed [2329.09k]                                  

automake/1.16.4@eriff/test: Copying sources to build folder
automake/1.16.4@eriff/test: Building your package in /home/eric/.conan/data/automake/1.16.4/eriff/test/build/258f6f05ca54813b3d7180161753df7f9ccf7e1d
automake/1.16.4@eriff/test: Generator txt created conanbuildinfo.txt
automake/1.16.4@eriff/test: Aggregating env generators
automake/1.16.4@eriff/test: Calling build()
automake/1.16.4@eriff/test: Calling:
 > /home/eric/.conan/data/automake/1.16.4/eriff/test/build/258f6f05ca54813b3d7180161753df7f9ccf7e1d/source_subfolder/configure '--datarootdir=/home/eric/.conan/data/automake/1.16.4/eriff/test/package/258f6f05ca54813b3d7180161753df7f9ccf7e1d/res' '--prefix=/home/eric/.conan/data/automake/1.16.4/eriff/test/package/258f6f05ca54813b3d7180161753df7f9ccf7e1d' '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libexecdir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' '--oldincludedir=${prefix}/include' --target=x86_64-linux-gnu
checking whether make supports nested variables... yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether ln -s works... yes
checking for perl... /usr/bin/perl
checking for tex... no
checking for yacc... no
checking for byacc... no
checking for bison... no
checking for lex... no
checking for flex... no
checking whether autoconf is installed... no
configure: error: Autoconf 2.65 or better is required.
    Please make sure it is installed and in your PATH.
```

Reply:
The log is lying to me since it says
```
autoconf/2.71: Appending PATH env var with : /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin
```

Reply:
Oh it looks like the issue comes from the `test_package`. If I remove `test_type = "build_requires"` then the library can be built (but the test package fails to build).
I'll investigate, but any comments are welcome.

Reply:
With this change I can build with a single profile and with pr:h=pr:b
```py
test_type = "build_requires", "requires"
```

Reply:
Continuing with my testing, I tried to build `libtool` using this freshly built `automake`. It doesn't work.
It had `automake` listed as `requirement`, so I had to conditionally add it as a build_req as well:
```py
    def requirements(self):
        self.requires("automake/1.16.4@eriff/test")

    def build_requirements(self):
        if hasattr(self, "settings_build"):
            self.build_requires("automake/1.16.4@eriff/test")
```
Whit those changes I'm able to build the package, but not the test_package
```
RROR: libtool/2.4.6@eriff/test (test package): Error in build() method, line 137
	self._build_autotools()
while calling '_build_autotools', line 43
	self.run("{} --install --verbose -Wall".format(os.environ["AUTORECONF"]), win_bash=tools.os_info.is_windows)
	KeyError: 'AUTORECONF'
```
So I resorted to my previous trick
```py
test_type = "build_requires", "requires"
```
And that made it happy.
Let's say we roll with this. I guess we would need similar fixes on a bunch of packages to enable building using pr:b=pr:h.
Is this a good fix or am I breaking something that I don't understand?

Reply:
I'm fine with this approach.
The test package needs autoconf+automake+m4 as a build requirement because the scripts expects some environment variables set:
some are overrides to support a portable autoconf/automake package,
others are overrides to support aclocal.

Does adding only `test_type = "build_requires"` work?
Because the test only uses autoconf/automake as a build requirement.

Reply:
No, it doesn't. And it is unclear to me why, but it looks like a bug on the conan client.
For some reason if I set `test_type = "build_requires"` then the package itself fails to build (not the test package). It complains about autoconf not being in the PATH, like if autoconf wasn't a build_req.

```
conan create recipes/automake/all automake/1.16.4@eriff/test -pr:b default -pr:h default
Exporting package recipe
automake/1.16.4@eriff/test exports: File 'conandata.yml' found. Exporting it...
automake/1.16.4@eriff/test exports: Copied 1 '.yml' file: conandata.yml
automake/1.16.4@eriff/test exports_sources: Copied 14 '.patch' files
automake/1.16.4@eriff/test: The stored package has not changed
automake/1.16.4@eriff/test: Exported revision: 6e8bf015f603ade25713733c68f78a8d
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=8
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

automake/1.16.4@eriff/test: Forced build from source
automake/1.16.4@eriff/test (test package): Installing package
Requirements
Packages
Build requirements
    autoconf/2.71 from 'conancenter' - Cache
    automake/1.16.4@eriff/test from local cache - Cache
    m4/1.4.19 from 'conancenter' - Cache
Build requirements packages
    autoconf/2.71:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    automake/1.16.4@eriff/test:258f6f05ca54813b3d7180161753df7f9ccf7e1d - Build
    m4/1.4.19:24647d9fe8ec489125dfbae4b3ebefaf7581674c - Cache

Installing (downloading, building) binaries...
m4/1.4.19: Already installed!
m4/1.4.19: Appending PATH environment variable: /home/eric/.conan/data/m4/1.4.19/_/_/package/24647d9fe8ec489125dfbae4b3ebefaf7581674c/bin
m4/1.4.19: Setting M4 environment variable: /home/eric/.conan/data/m4/1.4.19/_/_/package/24647d9fe8ec489125dfbae4b3ebefaf7581674c/bin/m4
autoconf/2.71: Already installed!
autoconf/2.71: Appending PATH env var with : /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin
autoconf/2.71: Setting AC_MACRODIR to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/share/autoconf
autoconf/2.71: Setting AUTOCONF to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/autoconf
autoconf/2.71: Setting AUTORECONF to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/autoreconf
autoconf/2.71: Setting AUTOHEADER to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/autoheader
autoconf/2.71: Setting AUTOM4TE to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/autom4te
autoconf/2.71: Setting AUTOM4TE_PERLLIBDIR to /home/eric/.conan/data/autoconf/2.71/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin/share/autoconf
automake/1.16.4@eriff/test: WARN: Build folder is dirty, removing it: /home/eric/.conan/data/automake/1.16.4/eriff/test/build/258f6f05ca54813b3d7180161753df7f9ccf7e1d
automake/1.16.4@eriff/test: Copying sources to build folder
automake/1.16.4@eriff/test: Building your package in /home/eric/.conan/data/automake/1.16.4/eriff/test/build/258f6f05ca54813b3d7180161753df7f9ccf7e1d
automake/1.16.4@eriff/test: Generator txt created conanbuildinfo.txt
automake/1.16.4@eriff/test: Aggregating env generators
automake/1.16.4@eriff/test: Calling build()

automake/1.16.4@eriff/test: Calling:
 > /home/eric/.conan/data/automake/1.16.4/eriff/test/build/258f6f05ca54813b3d7180161753df7f9ccf7e1d/source_subfolder/configure '--datarootdir=/home/eric/.conan/data/automake/1.16.4/eriff/test/package/258f6f05ca54813b3d7180161753df7f9ccf7e1d/res' '--prefix=/home/eric/.conan/data/automake/1.16.4/eriff/test/package/258f6f05ca54813b3d7180161753df7f9ccf7e1d' '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libexecdir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' '--oldincludedir=${prefix}/include' --target=x86_64-linux-gnu
checking whether make supports nested variables... yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether ln -s works... yes
checking for perl... /usr/bin/perl
checking for tex... no
checking for yacc... no
checking for byacc... no
checking for bison... no
checking for lex... no
checking for flex... no
checking whether autoconf is installed... no
configure: error: Autoconf 2.65 or better is required.
    Please make sure it is installed and in your PATH.
automake/1.16.4@eriff/test: 
```

Reply:
Thanks.
FYI, I cannot get autoconf to build with MSVC16 because autom4te fails due to `m4 --help` fails.
This is a problem with m4/1.4.19.
`m4 --help` exits with a non-zero return code.

Reply:
I see. But that is unrelated to this PR, right?

Reply:
Indeed. I just wanted to mention this because I found out about this when applying the changes of the autoconf recipe.

Reply:
Since you're playing with `m4`
Any reason why you did this on the `test_package`
```py
        if hasattr(self, "settings_build"):
            exe_suffix = ".exe" if self.settings.os == "Windows" else ""
            m4_bin = os.path.join(self.deps_cpp_info["m4"].rootpath, "bin", "m4" + exe_suffix)
        else:
            m4_bin = tools.get_env("M4")
            if m4_bin is None or not m4_bin.startswith(self.deps_cpp_info["m4"].rootpath):
                raise ConanException("M4 environment variable not set")
```
Instead of 
```py
test_type = "build_requires"
```
?

Reply:
I tried to test the host recipe, not the build requirement.
Environment variables of the host recipes are not set when using conan 2 profile builds.
Perhaps the test can become a pure `test_type = "build_requires"`.